Introduce `canonicalizeCandidates` on the internal Design System

RobinMalfait opened the pull request:
This PR introduces a new `canonicalizeCandidates` function on the internal Design System.

The big motivation to moving this to the core `tailwindcss` package is that we can use this in various places:

- The Raycast extension
- The VS Code extension / language server
- 3rd party tools that use the Tailwind CSS design system APIs

> This PR looks very big, but **I think it's best to go over the changes commit by commit**. Basically all of these steps already existed in the upgrade tool, but are now moved to our core `tailwindcss` package.

Here is a list of all the changes:
- Added a new `canonicalizeCandidates` function to the design system
- Moved various migration steps to the core package. I inlined them in the same file and because of that I noticed a specific pattern (more on this later).
- Moved `printCandidate` tests to the `tailwindcss` package
- Setup tests for `canonicalizeCandidates` based on the existing tests in the upgrade tool.

I noticed that all the migrations followed a specific pattern:

1. Parse the raw candidate into a `Candidate[]` AST
2. In a loop, try to migrate the `Candidate` to a new `Candidate` (this often handled both the `Candidate` and its `Variant[]`)
3. If something changed, print the new `Candidate` back to a string, and pass it to the next migration step.

While this makes sense in isolation, we are doing a lot of repeated work by parsing, modifying, and printing the candidate multiple times. This let me to introduce the `big refactor` commit. This changes the steps to:

1. Up front, parse the raw candidate into a `Candidate[]` _once_.
2. Strip the variants and the important marker from the candidate. This means that each migration step only has to deal with the base `utility` and not care about the variants or the important marker. We can re-attach these afterwards.
3. Instead of a `rawCandidate: string`, each migration step receives an actual `Candidate` object (or a `Variant` object).
4. I also split up the migration steps for the `Candidate` and the `Variant[]`.

All of this means that there is a lot less work that needs to be done. We can also cache results between migrations. So `[@media_print]:flex` and `[@media_print]:block` will result in `print:flex` and `print:block` respectively, but the `[@media_print]` part is only migrated once across both candidates.

One migration step relied on the `postcss-selector-parser` package to parse selectors and attribute selectors. I didn't want to introduce a package just for this, so instead used our own `SelectorParser` in the migration and wrote a small `AttributeSelectorParser` that can parse the attribute selector into a little data structure we can work with instead.

If we want, we can split this PR up into smaller pieces, but since the biggest chunk is moving existing code around, I think it's fairly doable to review as long as you go commit by commit.

---

With this new API, we can turn:
```
[
  'bg-red-500',
  'hover:bg-red-500',
  '[@media_print]:bg-red-500',
  'hover:[@media_print]:bg-red-500',
  'bg-red-500/100',
  'hover:bg-red-500/100',
  '[@media_print]:bg-red-500/100',
  'hover:[@media_print]:bg-red-500/100',
  'bg-[var(--color-red-500)]',
  'hover:bg-[var(--color-red-500)]',
  '[@media_print]:bg-[var(--color-red-500)]',
  'hover:[@media_print]:bg-[var(--color-red-500)]',
  'bg-[var(--color-red-500)]/100',
  'hover:bg-[var(--color-red-500)]/100',
  '[@media_print]:bg-[var(--color-red-500)]/100',
  'hover:[@media_print]:bg-[var(--color-red-500)]/100',
  'bg-(--color-red-500)',
  'hover:bg-(--color-red-500)',
  '[@media_print]:bg-(--color-red-500)',
  'hover:[@media_print]:bg-(--color-red-500)',
  'bg-(--color-red-500)/100',
  'hover:bg-(--color-red-500)/100',
  '[@media_print]:bg-(--color-red-500)/100',
  'hover:[@media_print]:bg-(--color-red-500)/100',
  'bg-[color:var(--color-red-500)]',
  'hover:bg-[color:var(--color-red-500)]',
  '[@media_print]:bg-[color:var(--color-red-500)]',
  'hover:[@media_print]:bg-[color:var(--color-red-500)]',
  'bg-[color:var(--color-red-500)]/100',
  'hover:bg-[color:var(--color-red-500)]/100',
  '[@media_print]:bg-[color:var(--color-red-500)]/100',
  'hover:[@media_print]:bg-[color:var(--color-red-500)]/100',
  'bg-(color:--color-red-500)',
  'hover:bg-(color:--color-red-500)',
  '[@media_print]:bg-(color:--color-red-500)',
  'hover:[@media_print]:bg-(color:--color-red-500)',
  'bg-(color:--color-red-500)/100',
  'hover:bg-(color:--color-red-500)/100',
  '[@media_print]:bg-(color:--color-red-500)/100',
  'hover:[@media_print]:bg-(color:--color-red-500)/100',
  '[background-color:var(--color-red-500)]',
  'hover:[background-color:var(--color-red-500)]',
  '[@media_print]:[background-color:var(--color-red-500)]',
  'hover:[@media_print]:[background-color:var(--color-red-500)]',
  '[background-color:var(--color-red-500)]/100',
  'hover:[background-color:var(--color-red-500)]/100',
  '[@media_print]:[background-color:var(--color-red-500)]/100',
  'hover:[@media_print]:[background-color:var(--color-red-500)]/100'
]
```

Into their canonicalized form:
```
[
  'bg-red-500',
  'hover:bg-red-500',
  'print:bg-red-500',
  'hover:print:bg-red-500'
]
```

The list is also unique, so we won't end up with `bg-red-500 bg-red-500` twice.

While the canonicalization itself is fairly fast, we still pay a **~1s** startup cost for some migrations (once, and cached for the entire lifetime of the design system). I would like to keep improving the performance and the kinds of migrations we do, but I think this is a good start.

The cost we pay is for:

1. Generating a full list of all possible utilities based on the `getClassList` suggestions API.
2. Generating a full list of all possible variants.

The canonicalization step for this list takes **~2.9ms** on my machine.

Just for fun, if you use the `getClassList` API for intellisense that generates all the suggestions and their modifiers, you get a list of **263788** classes. If you canonicalize all of these, it takes **~500ms** in total. So roughly **~1.9μs** per candidate.

This new API doesn't result in a performance difference for normal Tailwind CSS builds.

The other potential concern is file size of the package. The generated `tailwindcss.tgz` file changed like this:
```diff
- 684652 bytes (684.65 kB)
+ 749169 bytes (749.17 kB)
```
So the package increased by ~65 kB which I don't think is the end of the world, but it is important for the `@tailwindcss/browser` build which we don't want to grow unnecessarily. For this reason we remove some of the code for the design system conditionally such that you don't pay this cost in an environment where you will never need this API.

The `@tailwindcss/browser` build looks like this:
```shell
`dist/index.global.js 255.14 KB` (main)
`dist/index.global.js 272.61 KB` (before this change)
`dist/index.global.js 252.83 KB` (after this change, even smaller than on `main`)
```

RobinMalfait committed:
hoist regex

RobinMalfait committed:
expose `canonicalizeCandidates` on the design system

+ initial basic implementation

RobinMalfait committed:
remove async

While going through the migrations, I noticed that this was still marked
as async even though nothing is async about this. Simplified it to a
simple sync function instead.

RobinMalfait committed:
tmp: annotate migrations

This is likely temporary, but needed to know which migrations were sync
or async, and which migrations are an _actual_ migration from v3 → v4 or
which ones are mere optimizations in v4 land.

The v4 optimizations will be moved to the `canonicalizeCandidates` logic
straight on the design system.

RobinMalfait committed:
move `printCandidate` normalization tests to core

RobinMalfait committed:
canonicalize candidates at the end of upgrading your project

After all migrations happened, we run the canonicalization step to
perform the last bit of migrations

RobinMalfait committed:
move `migrate-bg-gradient` to core

RobinMalfait committed:
move `migrate-theme-to-var` to core

RobinMalfait committed:
cache the converter

We will likely clean this up later. But just trying to move things
aroudn first.

RobinMalfait committed:
only prefix variable that are not in `--theme(…)`

RobinMalfait committed:
move types to core

RobinMalfait committed:
move dimensions to core

RobinMalfait committed:
move signatures to core

RobinMalfait committed:
ensure `canonicalizeCandidates` returns a unique list

If you use `canonicalizeCandidates(['bg-red-500', 'bg-red-500/100'])`
they would both canonicalize to `bg-red-500`. No need to return
duplicates here.

RobinMalfait committed:
move `migrate-arbitrary-utilities` to core

Note: this commit on its own _will_ have some failing tests because the
`migrate-arbitrary-utilities.test.ts` file combined some other
migrations as well.

Eventually once those are ported over, will make the tests pass.

Another thing to notice is that we updated some tests. The more we
canonicalize candidates, the more tests will need upgrading to an even
more canonical form.

RobinMalfait committed:
move `migrate-bare-utilities` to core

RobinMalfait committed:
move `migrate-deprecated-utilities` to core

RobinMalfait committed:
move `replaceObject` to core

RobinMalfait committed:
move `migrate-arbitrary-variants` to core

RobinMalfait committed:
move `migrate-drop-unnecessary-data-types` to core

RobinMalfait committed:
move `migrate-arbitrary-value-to-bare-value` to core

RobinMalfait committed:
move `migrate-optimize-modifier` to core

RobinMalfait committed:
remove `!` from test case

This test setup already tests various cases including prefixes and
important strategies.

RobinMalfait committed:
handle `&` and `*` in selector parser as standalone selector

We were also converting the current charcode back to a string, but we
already had that information in `input[i]`

RobinMalfait committed:
move parts of `migrate-modernize-arbitrary-values` to core

The part where we upgrade old v3 syntax like `group-[]:` still exists in
the `@tailwindcss/upgrade` package.

We were relying on `postcss-selector-parser`, but the copied part is now
using our own `SelectorParser`.

I also added a temporary attribute selector parser to make everything
pass. Don't look at the code for that _too_ much because we will swap it
out in a later commit 😅

RobinMalfait committed:
ensure we canonicalize at the end

A bunch of migrations were moved to `canonicalizeCandidates`, so we
still have to run that at the end.

Also used:
```css
@theme {
  --*: initial;
}
```

To make the design system a bit faster.

RobinMalfait committed:
big refactor

Note: we made sure that all the tests passed before this commit, and we
also made sure to not touch any of the tests as part of this commit
ensuring that everything still works as expected. Aka, an actual
refactor without behavior change.

This is a big commit and a big refactor but that's mainly because there
are some re-indents that are happening. Will maybe split up this commit
but then it won't be an atomic (working) commit anymore.

One thing I noticed is that all our "migrations" essentially parsed a
raw candidate, did some work and then stringified the candidate AST
again.

That's one of the reasons I inlined the migrations in previous commits
in a single file. Looking at the file now, it's a little bit silly that
we parse and print over and over again.

Parsing a raw candidate also results in an an array of `Candidate` AST
nodes, so we always had a loop going on.

Another thing I noticed is that often we need to work with a "base"
candidate, that's essentially the candidate without the variants and
without the important flag.

That's where this refactor comes in:

1. We parse each candidate _once_
2. When there are variants, we canonicalize each variant separately and
   store the results in a separate cache.
3. When there are variants _or_ the important flag is present, we
   canonicalize the base utility separately and later re-attach the
   variants and important flag at the end. This means that not a single
   canonicalization step has to deal with variants or important flag.
4. Since we only want to parse the candidate once on the outside, we
   will pass a cloned `Candidate` to each canonicalize step to prevent
   parsing and printing over and over again.
5. Only at the end will we re-print the `Candidate`

This process has a few benefits:

1. We are handling candidates and variants separately and caching them
   separately. This means that we can increase cache hits because
   variants from candidate A and candidate B can be re-used. E.g.:
   `[@media_print]:flex` and `[@media_print]:underline` will map to
   `print:flex` and `print:underline` where `[@media_print]` is only
   handled once.
2. This also means that we could simplify some canonicalization steps
   because they don't have to worry about variants and the important
   flag.
3. There is no parsing & looping over the parsed candidates array going
   on.

RobinMalfait committed:
drop unnecessary calls

The the scenario of creating signatures for a component, we always want
to respect the important flag if it is given. Therefore I think that we
can drop the `designSystem.compileAstNodes(candidate,
CompileAstFlags.None)` which saves us some time.

Also removed the `structuredClone` call. Whenever we handle `@apply`
which essentially happened the step before, we already start from a
cloned node internally. This call just meant we were doing it again.

RobinMalfait committed:
only stringify the value when something changed

RobinMalfait committed:
remove migration annotations

RobinMalfait committed:
move `selector-parser` from `./src/compat` to just `./src`

RobinMalfait committed:
add attribute selector parser

RobinMalfait committed:
use dedicated `AttributeSelectorParser`

This implementation is also 2-3x faster than the original one.

```
AttributeSelectorParser.parse - src/attribute-selector-parser.bench.ts > parsing
  2.24x faster than REGEX.test(…)
  2.59x faster than ….match(REGEX)
  2.84x faster than parseAttributeSelector
```

RobinMalfait commented on packages/tailwindcss/src/canonicalize-candidates.test.ts:
```diff
@@ -97,7 +97,7 @@ describe.each([['default'], ['with-variant'], ['important'], ['prefix']])('%s',
     /// theme(…) to `var(…)`
     // Keep candidates that don't contain `theme(…)` or `theme(…, …)`
-    ['[color:red]', '[color:red]'],
+    ['[color:red]', 'text-[red]'],
```
So you will notice that tests changed in this commit. This is because we were testing each migration in isolation before. But now we're testing the entire `canonicalizeCandidates` function at once.

It just happens to be that this is an even _more_ canonicalized compared to before.

RobinMalfait commented on packages/tailwindcss/src/compat/selector-parser.test.ts:
```diff
@@ -131,6 +131,34 @@ describe('parse', () => {
+  })
+
+  it('parses nesting selector after an attribute selector', () => {
+    expect(parse('[data-foo]&')).toEqual([
```
While converting from the `postcss-selector-parser` setup to our own `SelectorParser`, I noticed that this case wasn't covered properly and broke some tests.

RobinMalfait commented on packages/tailwindcss/src/canonicalize-candidates.ts:
```diff
@@ -1110,6 +1112,516 @@ function* tryValueReplacements(
+  End,
+}
+
+function parseAttributeSelector(value: string) {
```
Don't read too much into this function. This was a temporary replacement for the attribute selector parser. I added a proper commit with an implementation + tests + benchmark. Just needed this to make the tests pass.

RobinMalfait commented on packages/tailwindcss/src/canonicalize-candidates.ts:
```diff
@@ -25,65 +25,146 @@ import * as ValueParser from './value-parser'
 
 export function canonicalizeCandidates(ds: DesignSystem, candidates: string[]): string[] {
   let result = new Set<string>()
+  let cache = canonicalizeCandidateCache.get(ds)
```
Might be useful to look at this commit with `?w=1` as well due to the re-indents.

thecrypticace commented:
Yeah core isn't 3.5MB right now. It's about 680KB (which is still a lot but that's because CJS + ESM duplication).

The build process doesn't clear the `dist` folder so it probably got a bunch of chunks from previous builds bundled into the .tgz folder on Robin's local machine.

RobinMalfait committed:
remove intellisense features from `@tailwindcss/browser`

RobinMalfait commented:
> Yeah core isn't 3.5MB right now. It's about 680KB (which is still a lot but that's because CJS + ESM duplication).
> 
> The build process doesn't clear the `dist` folder so it probably got a bunch of chunks from previous builds bundled into the .tgz folder on Robin's local machine.

Yep, that was the reason why! I'm not _that_ stupid and I did clear them manually before. But I'm stupid enough not to realise that Turborepo just put them all back when running a build as if nothing ever happened. 😅

thecrypticace commented on packages/tailwindcss/src/compat/selector-parser.ts:
```diff
@@ -369,7 +371,7 @@ export function parse(input: string) {
             ast.push(node)
           }
         }
-        buffer = String.fromCharCode(currentChar)
```
I was thinking we did this in purpose for perf reasons but it might not actually make a difference compared to everything else.

thecrypticace commented on packages/tailwindcss/src/canonicalize-candidates.ts:
```diff
@@ -0,0 +1,1510 @@
+    }
+
+    // Update the candidate with the new value
+    replaceObject(candidate, replacementCandidate)
```
I think all the uses of `replaceObject(candidate, replacementCandidate); return replacementCandidate` can be just `return replacementCandidate` right?

It'd be cool if we could get rid of `replaceObject` entirely but that's probably not realistic.

thecrypticace commented on packages/tailwindcss/src/attribute-selector-parser.ts:
Does it make sense to merge this with the main selector parser or nah?

RobinMalfait replied:
I didn't because we don't _need_ it anywhere else. But we can make it part of the SelectorParser and introduce a new `attribute` kind 🤔 

I think the `postcss-selector-parser` actually has a `selector` kind that is just a group, each node is `class`, `id`, `attribute` etc.

The only other spot in core where we use the selector parser is inside of the compat plugin API

thecrypticace replied:
I'm kinda indifferent — it just feels odd to have a "selector parser" and a separate "attribute selector parser". Reads like the selector parser doesn't handle attributes at all. (it does just doesn't split out their "parts").

Maybe @philipp-spiess can weigh in on what he thinks.

philipp-spiess replied:
So, right now the selector parser just skips over attribute selectors until the closing `]`, right? Seems like in the future we might merge those two but right now if there's no overlap I don't think we have to bother. In fact, using the attribute selector parser from within the selector parser in the attribute branch should be fairly trivial now. I'd say we leave it like this for now but maybe add a comment here explaining why we don't do more work yet and where that work can be found: https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/compat/selector-parser.ts#L378-L406

thecrypticace reviewed (approved):
Overall things look good 💯 

Massive PR but will be awesome to have this stuff in core. If we can speed up the initialization step enough I'll be able to make this a "default on" diagnostic for intellisense which would be pretty cool.

RobinMalfait committed:
immediately return replacement candidates

RobinMalfait committed:
remove `@property` when computing a signature

These are globally inserted and don't super matter, at least I don't
think so and it will free up a lot of memory.

~213 MB → ~154 MB

itsdouges commented on packages/@tailwindcss-browser/tsup.config.ts:
```diff
@@ -13,4 +13,20 @@ export default defineConfig({
+        build.onLoad({ filter: /intellisense.ts$/ }, () => {
+          return {
+            contents: `
+              export function getClassList() { return [] }
```
Does this mean these APIs aren't available when running the browser version?

thecrypticace replied:
These APIs are internal and were already inaccessible in the `@tailwindcss/browser` package. They were on an (i.e. not tree-shakable) which meant they still took up some space. This replaces those APIs so they take up much less space since they can't be used anyway.

basically this is just a code-size reduction and doesn't eliminate any actual functionality.

itsdouges replied:
Ah that must be it great, thanks mate

itsdouges commented on packages/tailwindcss/src/attribute-selector-parser.bench.ts:
```diff
@@ -0,0 +1,35 @@
+  '[data-state*=expanded i]',
+]
+
+const ATTRIBUTE_REGEX =
+  /\[\s*(?<attribute>[a-zA-Z_-][a-zA-Z0-9_-]*)\s*((?<operator>[*|~^$]?=)\s*(?<quote>['"])?\s*(?<value>.*?)\4\s*(?<sensitivity>[is])?\s*)?\]/
```
Are you balancing correctness, bundle size and performance? There could always be work done on a proper parser here but of course comes with more code as a cost

RobinMalfait committed:
add note when parsing the attribute selector as a whole
